_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+        │  └╴← (uint256)
+        ├╴call ERC20::balanceOf(address)(address) (src/Solidity.t.sol:96)
+        │  └╴← (uint256)
+        ├╴call ERC20::transfer(address,uint256)(address, uint256) (src/Solidity.t.sol:98)
+        │  ├╴Transfer(uint256) (src/Solidity.t.sol:67)
+        │  └╴← (bool)
+        ├╴call ERC20::balanceOf(address)(address) (src/Solidity.t.sol:101)
+        │  └╴← (uint256)
+        ├╴log_bytes32(bytes32) (lib/ds-test/src/test.sol:108)
+        ├╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:109)
+        ├╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:110)
+        ├╴call ERC20::balanceOf(address)(address) (src/Solidity.t.sol:104)
+        │  └╴← (uint256)
+        ├╴log_bytes32(bytes32) (lib/ds-test/src/test.sol:108)
+        ├╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:109)
+        └╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:110)
+```
+
+Looking into the output, we can see that this represents the case where `dst` is the same as the
+sender (in this case the test contract).
+
+In this case the counter example found doesn't represent a bug in the implementation of `transfer`,
+but rather shows that our understanding of the expected behaviour was flawed: an exhaustive
+description of the behaviour of `transfer` must take self transfers into account. This kind of
+situation is common when applying formal methods, where we are forced to consider all possible edge
+cases.
+
+It is also worth noting that fuzzing would be very unlikely to catch this edge case: there are 2^20
+possible addresses, and the chance that a randomly generated address would match the address of the
+test contract is miniscule. You can try it out yourself by renaming the `prove_transfer` method to
+`test_transfer` and seeing if a counter example is found.
+
+A test for `transfer` that takes self-transfers into account could look like this:
+
+```solidity
+function prove_transfer(address dst, uint amt) public {
+    log_named_address("this", address(this));
+
+    uint preBalThis = token.balanceOf(address(this));
+    uint preBalDst  = token.balanceOf(dst);
+
+    token.transfer(dst, amt);
+
+    // no change for self-transfer
+    uint delta = dst == address(this) ? 0 : amt;
+
+    // balance of this has been reduced by `delta`
+    assertEq(sub(preBalThis, token.balanceOf(address(this))), delta);
+
+    // balance of dst has been increased by `delta`
+    assertEq(sub(token.balanceOf(dst), preBalDst), delta);
+}
+```
+
+## Execution Environment And Limits to Proof
+
+In order to understand the limits of the proofs that it is possible to produce using this framework,
+an understanding of the environment in which they are run is essential:
+
+- All variables in the environment (e.g. caller, gas, timestamp) remain concrete
+- All storage slots are set to zero at the beginning of the tests (meanings tests are effectively run against an empty blockchain)
```

Comment:
well, they are set to whatever happened in the setup, which may include rpc requests...